### PR TITLE
ARROW-1774: [C++] Add Array::View()

### PR DESCRIPTION
**Pull request by pitrou**

### CODE REVIEW

**Comment by pitrou** [comment]
@wesm I'm unsure what to do with offsets here. Perhaps as a first approach, we should raise on non-zero input offsets?

**Comment by pitrou** [comment]
The approach in this PR is generic and powerful, as it allows non-trivial views (see unit tests), however it may also produce incorrect results in some select cases (e.g. when viewing a fixed-size list as a primitive array). Should we be concerned with that or just have the user be careful?

File: cpp/src/arrow/type.h [hunk 1]
```diff
@@ -147,6 +147,15 @@ struct Type {
   };
 };
 
+struct ARROW_EXPORT DataTypeLayout {
+  // The bit width for each buffer in this DataType's representation
+  // (-1 if the item size for a given buffer is unknown or variable,
```
Comment (fsaintjacques) [inline cpp/src/arrow/type.h]: Make `-1` a constexpr.

File: cpp/src/arrow/type.h [hunk 2]
```diff
@@ -147,6 +147,15 @@ struct Type {
...
+  //  0 if the buffer is always null).
+  // Child types are not included, they should be inspected separately.
+  std::vector<int64_t> bit_widths;
+  bool has_dictionary;
+};
```
Comment (fsaintjacques) [inline cpp/src/arrow/type.h]: Why should children be excluded? It's also missing the union mode which dictates how to interpret the children arrays.

File: cpp/src/arrow/type.h [hunk 3]
```diff
@@ -147,6 +147,15 @@ struct Type {
   };
 };
 
+struct ARROW_EXPORT DataTypeLayout {
+  // The bit width for each buffer in this DataType's representation
+  // (-1 if the item size for a given buffer is unknown or variable,
+  //  0 if the buffer is always null).
+  // Child types are not included, they should be inspected separately.
+  std::vector<int64_t> bit_widths;
+  bool has_dictionary;
```
Comment (fsaintjacques) [inline cpp/src/arrow/type.h]: Can you specify a default?

File: cpp/src/arrow/array.cc [hunk 4]
```diff
@@ -831,6 +831,202 @@ Status DictionaryArray::Transpose(MemoryPool* pool, const std::shared_ptr<DataTy
...
+      DCHECK_GT(in_data_item->buffers.size(), in_buffer_idx);
+      out_buffers.push_back(in_data_item->buffers[in_buffer_idx]);
+      out_length = in_data_item->length;
+      out_null_count = in_data_item->null_count;
+      ++in_buffer_idx;
+      AdjustInputPointer();
+    } else {
+      // No null bitmap in input, append no-nulls bitmap
+      out_buffers.push_back(nullptr);
+      out_null_count = 0;
```
Comment (wesm) [inline cpp/src/arrow/array.cc]: Does this occur anywhere now?

**Review by wesm — COMMENTED** [review]
This does look very flexible; the implementation takes some effort to reason through but it's compact so that is nice =) 

Yes, I think that raising for non-zero `ArrayData::offset` as a first pass seems OK (assuming that's what you meant). Have you thought of cases where passing through the offset as with the length would cause issues?

File: cpp/src/arrow/type.h (hunk 2 above)
Comment (pitrou) [reply cpp/src/arrow/type.h]: For several reasons:
1. the caller may not be interested directly in children
2. handling the information for children would make the structure self-referential
3. the caller may want to do other things while recursing, or recurse selectively, etc.

As for the union mode, this struct only describes the physical layout, not how to interpret arrays. How to interpret arrays is embodied in the DataType itself.

File: cpp/src/arrow/type.h (hunk 3 above)
Comment (pitrou) [reply cpp/src/arrow/type.h]: I'd like to but C++ disables default constructors when you add a default initializer to a member field :-/

File: cpp/src/arrow/array-view-test.cc [hunk 5]
```diff
@@ -0,0 +1,377 @@
...
+  // With nulls
+  arr = ArrayFromJSON(ty1, "[[0], null, [-1]]");
+  expected = ArrayFromJSON(uint16(), "[0, null, 65535]");
+  //   CheckView(arr, expected);  // XXX currently fails
+  CheckView(expected, arr);
+
+  // With nested nulls => fails
+  arr = ArrayFromJSON(ty1, "[[0], [null], [-1]]");
+  CheckViewFails(arr, uint16());
+}
```
[the remark this answers is not in the document]
Comment (pitrou) [reply cpp/src/arrow/array-view-test.cc]: I could remove the test and the fact that this works would then be an implementation detail. What do you think?

[the remark this answers is not in the document]
Comment (wesm) [reply cpp/src/arrow/array-view-test.cc]: I think it's actually kind of useful, to be able to View a primitive as a struct, so I'm OK to leave it.

**Comment by wesm** [comment]
@pitrou I'd be OK to merge this with an error on non-zero offsets for the time being, then we can address that as a follow up JIRA

File: cpp/src/arrow/array.cc (hunk 4 above)
Comment (pitrou) [reply cpp/src/arrow/array.cc]: Yes, when viewing structs as structs of different shapes (for example `struct<a: struct<b: int8>, d: int32>` as `struct<a: uint8, b: struct<b: float>>`).

**Review by wesm — APPROVED** [review]
+1

**Comment by wesm** [comment]
I opened https://issues.apache.org/jira/browse/ARROW-5558 about non-zero offsets

**Pull request closed without merging**